### ISSUE TITLE
ctc_loss with best align path

### ISSUE THREAD

**Issue by xushenkun** [tensorflow/tensorflow#15028]
The tf.nn.ctc_loss function only returns negative log probabilities according to truth labels for the sequence's best align path. 
But we also need this best align path in some situations.
Could you please supply this feature, thanks!
------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MINGW64_NT-6.1 xsk-PC 2.6.0(0.304/5/3) 2016-09-07 20:45 x86_64 Msys
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.4.0
- **Python version**: 3.6.0
- **Bazel version (if compiling from source)**: None
- **GCC/Compiler version (if compiling from source)**: None
- **CUDA/cuDNN version**: None
- **GPU model and memory**: None
- **Exact command to reproduce**: None

**Comment by carlthome** [comment]
Use `tf.nn.ctc_beam_search_decoder`?

**Comment by xushenkun** [comment]
ctc_beam_search_decoder and ctc_greedy_decoder can only make predications, but can't get the best alignment path according to the target.

**Comment by xushenkun** [comment]
According to "https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/util/ctc/ctc_loss_calculator.cc", I think the best alignment and probs may be as below and this is not a waste of resources during calculating ctc loss.

best_align = argmax(log_alpha, axis=0) 
blank_pos = where(best_align % 2 == 0)
target_pos = where(best_align % 2 != 0)
best_align[target_pos] = target[best_align[target_pos] // 2]
best_align[blank_pos] = blank_index_
best_prob = logit[best_align, range(T)]

**Comment by drpngx** [comment]
If you count the blanks, you should have the alignment.

**Comment by xushenkun** [comment]
@drpngx Could you please explain the meaning of 'count the blanks' more detail. After I used Viterbi with ctc to find the most probable path, which is not the true align path. I think ctc loss just make sure the sum of probs is biggest for all probable paths, but the prob of true align path may be very small among the candidate paths. And we can't get the real align path in fact according to ctc loss.

**Comment by carlthome** [comment]
Could you clarify the difference between _the most probable path_ and _the true align path_? Do you mean you want the output sequence without collapsed repeats and including blanks so it's of the same time resolution as your input?

**Comment by drpngx** [comment]
@carlthome , @xushenkun is saying that the decode does not allow you to constrain the path for alignment, and the loss function computes the full forward probability (in the log semiring, not max). For instance if you computed the top1 from beam search and used the `ctc_loss`, then you'd get a probability that's greater than or equal t the one returned by the decoder.

@xushenkun you'd need to either have an option to the ctc loss to use max instead of logsum, or have another version of the decoder that takes in the sequence (possibly without epsilons).

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by ebrevdo** [comment]
+1; if anyone is interested in implementing either of @drpngx's suggested options, should send us a PR.  it would require a new op (not modifying the existing CTC Loss / Decoders) because it would change the signature.  The new op can be exposed in tf.contrib.

**Comment by zjmwqx** [comment]
> @drpngx Could you please explain the meaning of 'count the blanks' more detail. After I used Viterbi with ctc to find the most probable path, which is not the true align path. I think ctc loss just make sure the sum of probs is biggest for all probable paths, but the prob of true align path may be very small among the candidate paths. And we can't get the real align path in fact according to ctc loss.

have you solved this problem? I am trying to algin the ctc decoded output to the raw timeline, can you help me?

**Comment by sushreebarsa** [comment]
Hi @xushenkun! TF v1.x is not actively supported, could you try with the latest TF v2.4 or later and let us know if the issue still persists?
Please provide the standalone code to reproduce the issue in latest TF versions if you still face the issue. Thank you!
